packages/liveblocks-devtools/src/devtools/tabs/yjs/to-yjs-tree-node.ts

```diff
@@ -0,0 +1,270 @@
+  id: string;
+  key: string;
+  payload: string;
+};
+
+export type YContentEmbedTreeNode = {
+  type: YContentEmbedNode["type"];
+  id: string;
+  key: string;
+  payload: Object;
+};
+
+export type YContentBinaryTreeNode = {
+  type: YContentBinaryNode["type"];
+  id: string;
+  key: string;
+  payload: Uint8Array;
+};
+
+export type YContentFormatTreeNode = {
+  type: YContentFormatNode["type"];
+  id: string;
+  key: string;
+  payload: Object;
+};
+
+export type YContentAnyTreeNode = {
+  type: YContentAnyNode["type"];
+  id: string;
+  key: string;
+  payload: (Number | Object | Boolean | Array<unknown> | String)[];
+};
+
+export type YContentTreeNode =
+  | YContentStringTreeNode
+  | YContentEmbedTreeNode
+  | YContentBinaryTreeNode
+  | YContentFormatTreeNode
+  | YContentAnyTreeNode;
+
+export type YTreeNode =
+  | YDocTreeNode
+  | YAbstractTypeTreeNode
+  | YContentTreeNode
+  | YTopAbstractTypeTreeNode;
+
+export function toTreeYNode(
+  node: YDocNode,
+  key?: string,
+  parentId?: string
+): YDocTreeNode;
+
+export function toTreeYNode(
+  node: YTopAbstractTypeNode,
+  key?: string,
+  parentId?: string
+): YTopAbstractTypeTreeNode;
+
+export function toTreeYNode(
+  node: YAbstractTypeNode,
+  key?: string,
+  parentId?: string
+): YAbstractTypeTreeNode;
+
+export function toTreeYNode(
+  node: YContentStringNode,
+  key?: string,
+  parentId?: string
+): YContentStringTreeNode;
+
+export function toTreeYNode(
+  node: YContentEmbedNode,
+  key?: string,
+  parentId?: string
+): YContentEmbedTreeNode;
+
+export function toTreeYNode(
+  node: YContentBinaryNode,
+  key?: string,
+  parentId?: string
+): YContentBinaryTreeNode;
+
+export function toTreeYNode(
+  node: YContentFormatNode,
+  key?: string,
+  parentId?: string
+): YContentFormatTreeNode;
+
+export function toTreeYNode(
+  node: YContentAnyNode,
+  key?: string,
+  parentId?: string
+): YContentAnyTreeNode;
+
+export function toTreeYNode(
+  node: YDocNode | YTopAbstractTypeNode | YAbstractTypeNode | YContentNode,
+  key = "0",
+  parentId = ""
+):
+  | YDocTreeNode
+  | YTopAbstractTypeTreeNode
+  | YAbstractTypeTreeNode
+  | YContentTreeNode {
+  const id = createTreeNodeId(key, parentId);
+
+  switch (node.type) {
+    case "Y.Doc":
+      return {
+        type: node.type,
+        id,
+        guid: node.guid,
+        key,
+        payload: Object.entries(node.data).map(([key, value]) =>
+          toTreeYNode(value, key, id)
+        ),
+      };
+    case "Y.AbstractType":
+      return {
+        type: node.type,
+        id,
+        key,
+        payload: Object.entries(node.data).map(([key, value]) => {
+          switch (value.type) {
+            case "Y.Doc":
+              return toTreeYNode(value, key, id);
+            case "Y.Map":
+            case "Y.Array":
+            case "Y.XmlFragment":
+            case "Y.XmlElement":
+            case "Y.Text":
+            case "Y.XmlText":
+              return toTreeYNode(value, key, id);
```

Comment:
if I'm reading this correctly, all of these switch items result to toTreeYNode, so would it be better to just do like `['Y.Map', etc...].includes(value.type) return toTreeYNode(value, key, id)`

Reply:
I do not like it either but TypeScript wasn't happy when I did that : (